decoder.go

```diff
@@ -99,44 +122,54 @@ func newDecoder(codecID C.enum_AVCodecID, logger logging.Logger) (*decoder, erro
 		return nil, errors.New("avcodec_open2() failed")
 	}
 
-	srcFrame := C.av_frame_alloc()
+	srcFrame, err := allocateAVFrame()
+	if err != nil {
+		return nil, errors.Errorf("AV Frame allocation error during decoder init: %v", err)
+	}
 	if srcFrame == nil {
 		C.avcodec_close(codecCtx)
 		return nil, errors.New("av_frame_alloc() failed")
 	}
 
 	return &decoder{
-		logger:   logger,
-		codecCtx: codecCtx,
-		srcFrame: srcFrame,
+		logger:      logger,
+		codecCtx:    codecCtx,
+		src:         srcFrame,
+		avFramePool: avFramePool,
 	}, nil
 }
 
 // newH264Decoder creates a new H264 decoder.
-func newH264Decoder(logger logging.Logger) (*decoder, error) {
-	return newDecoder(C.AV_CODEC_ID_H264, logger)
+func newH264Decoder(avFramePool *sync.Pool, logger logging.Logger) (*decoder, error) {
+	return newDecoder(C.AV_CODEC_ID_H264, avFramePool, logger)
 }
 
 // newH265Decoder creates a new H265 decoder.
-func newH265Decoder(logger logging.Logger) (*decoder, error) {
-	return newDecoder(C.AV_CODEC_ID_H265, logger)
+func newH265Decoder(avFramePool *sync.Pool, logger logging.Logger) (*decoder, error) {
+	return newDecoder(C.AV_CODEC_ID_H265, avFramePool, logger)
 }
 
 // close closes the decoder.
 func (d *decoder) close() {
-	if d.dstFrame != nil {
-		C.av_frame_free(&d.dstFrame)
+	if d.dst != nil {
+		d.dst = nil
 	}
 
 	if d.swsCtx != nil {
 		C.sws_freeContext(d.swsCtx)
 	}
 
-	C.av_frame_free(&d.srcFrame)
-	C.avcodec_close(d.codecCtx)
+	if d.src != nil {
+		d.src = nil
+	}
+
+	if d.codecCtx != nil {
+		C.avcodec_close(d.codecCtx)
+		d.codecCtx = nil
+	}
 }
 
-func (d *decoder) decode(nalu []byte) (image.Image, error) {
+func (d *decoder) decode(nalu []byte) (*imageAndPoolItem, error) {
```

Comment:
why are we we returning a pointer here? 
this is the definition of imageAndPoolItem:
```
	imageAndPoolItem struct {
		img      image.Image
		poolItem *avFrameWrapper
	}
```

it has 2 pointers (image.Image is an interface so it's a pointer). Why are we returning a pointer to a struct that contains pointers? Lets just return the struct.

Reply:
Updated. It is now just a struct returned